### PR TITLE
x-wso2-owner extension addition

### LINKED ISSUE #937: Microgateway Analytics data publishing through a gRPC service
### Describe your problem(s)

Currently Microgateway is using file upload methodology to publish analytics data to APIM analytics.
It writes analytics data to a zip file (file rotating task) and uploads it to APIM Analytics server (file uploading task). We can specify time intervals for above mentioned tasks.  But below are the limitations of existing method.

- Creation of large file sizes.
- Size of the file depending on transactions per a second.
- Inability to view real time analytics results.
- Incapability to handle large request counts.

### Describe your solution

We can override above limitations by using a light weight, effective and reliable communication mechanism. Using a gRPC service to publish analytics data satisfies above requirements. Also we can use HTTPS connection with gRPC to securely publish analytics data. Below list demonstrates some extra advantages of using a gRPC service. 

- Having small payload content (lesser overhead for communication).
- Ability to provide faster, efficient and reliable communication compared to ordinary RPC (Remote Procedure Calls).
- Real-time analytics data publishing. 

### How will you implement it

Following proto file will be used to define the message structure as well as the RPC method.

```
syntax = "proto3";

option java_multiple_files = true;
option java_package = "org.wso2.analytics.mgw.grpc.service";
import "google/protobuf/empty.proto";

message AnalyticsStreamMessage {
    string messageStreamName = 1;
    string meta_clientType = 2;
    string applicationConsumerKey = 3;
    string applicationName = 4;
    string applicationId = 5;
    string applicationOwner = 6;
    string apiContext = 7;
    string apiName = 8;
    string apiVersion = 9;
    string apiResourcePath = 10;
    string apiResourceTemplate = 11;
    string apiMethod = 12;
    string apiCreator = 13;
    string apiCreatorTenantDomain = 14;
    string apiTier = 15;
    string apiHostname = 16;
    string username = 17;
    string userTenantDomain = 18;
    string userIp = 19;
    string userAgent = 20;
    int64 requestTimestamp  = 21;
    bool throttledOut = 22;
    int64 responseTime = 23;
    int64 serviceTime  = 24;
    int64 backendTime = 25;
    bool responseCacheHit = 26;
    int64 responseSize = 27;
    string protocol = 28;
    int32 responseCode = 29;
    string destination = 30;
    int64 securityLatency = 31;
    int64 throttlingLatency = 32;
    int64 requestMedLat = 33;
    int64 responseMedLat = 34;
    int64 backendLatency = 35;
    int64 otherLatency = 36;
    string gatewayType = 37;
    string label = 38;
    //message attributes unique to throttle stream
    string subscriber = 39;
    string throttledOutReason = 40;
    int64 throttledOutTimestamp = 41;
    string hostname = 42;
   //message attributes unique to fault stream
    string errorCode = 43;
    string errorMessage = 44;
}
service AnalyticsSendService {
    rpc sendAnalytics(stream AnalyticsStreamMessage) returns(google.protobuf.Empty);
}
```

Suggested implementation publishes analytics data relevant to each API invocation. It could be a successful , throttled-out or a fault invocation.  Based on the invocation result, analytics data get created on behalf of it. Then gRPC message generation happens according to the relevant attributes defined in the proto file and invocation result. Below are the method names used to generate proto file's AnalyticsStreamMessage. 
- createResponseMessage
- createThrottleMessage
- createFaultMessage

Apart from message generation, above methods assign relevant stream name to the message based on the result.  
Then it invokes ballerina gRPC client and submits data to APIM Analytic's Siddhi Application over a **HTTPS streaming** connection in **asynchronous** manner.

image

---
#### Related Issues:

N/A
#### Suggested Labels:

N/A
#### Suggested Assignees:

N/A

### PR DESCRIPTION
**Pull request by 7slcode7**
### Purpose

To publish analytics data similar to APIM API invokes. Before adding this extension there were some mismatches in the APIM Analytics dashboard values (API version usage, total API usage etc).

- Displays analytics data correctly in APIM Analytics dashboard tables
- Updates API usage pie chart correctly

Also adds new wso2 carbon certificate under the name apim300.

### Issues

Fixes #937 
- Total API invokes for APIM and Microgateway not updated properly in the API usage pie chart (Microgateway API invokes not included in the total count).
- Also API invoke count listed as "unknown" field value under a separate row. (For imported APIs from APIM)

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

- Ubuntu 18.04.3 LTS
- Ballerina Version (1.1.0)
- Java open JDK version "11.0.5"
- wso2am-analytics-3.1.0-alpha
- wso2am-3.1.0-alpha

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-core/src/main/ballerina/src/gateway/analytics/responsetime_util.bal [hunk 1]
```diff
@@ -99,8 +98,10 @@ public function generateRequestResponseExecutionDataEvent(http:Response response
     }
     APIConfiguration? apiConfiguration = apiConfigAnnotationMap[context.getServiceName()];
     if (apiConfiguration is APIConfiguration) {
-        if (!stringutils:equalsIgnoreCase("", <string>apiConfiguration.publisher)) {
+        if (!stringutils:equalsIgnoreCase("", <string>apiConfiguration.publisher) && stringutils:equalsIgnoreCase("", <string>requestResponseExecutionDTO.apiCreator)) {
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/analytics/responsetime_util.bal]: 120 character length exceeds.

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java [hunk 2]
```diff
@@ -289,6 +300,10 @@ public static void setAdditionalConfigsDevFirst(ExtendedAPI api, OpenAPI openAPI
 
         setMgwAPISecurityAndScopes(api, openAPI);
         api.setSpecificBasepath(openAPI.getExtensions().get(OpenAPIConstants.BASEPATH).toString());
+        //assigns x-wso2-owner value to API provider
+        if (openAPI.getExtensions().get(OpenAPIConstants.API_OWNER) != null) {
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java]:
```suggestion
        if (openAPI.getExtensions().containsKey(OpenAPIConstants.API_OWNER)) {
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/analytics/responsetime_util.bal (hunk 1 above)
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/analytics/responsetime_util.bal]:
```suggestion
        if (!stringutils:equalsIgnoreCase("", <string>apiConfiguration.publisher) 
            && stringutils:equalsIgnoreCase("", <string>requestResponseExecutionDTO.apiCreator)) {
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/utils/analytics_util.bal [hunk 3]
```diff
@@ -35,7 +36,13 @@ function populateThrottleAnalyticsDTO(http:FilterContext context) returns (Throt
     APIConfiguration? apiConfiguration = apiConfigAnnotationMap[context.getServiceName()];
     if (apiConfiguration is APIConfiguration) {
         eventDto.apiVersion = apiConfiguration.apiVersion;
-    }
+        if (!stringutils:equalsIgnoreCase("", <string>apiConfiguration.publisher)) {
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/utils/analytics_util.bal]: duplicated code in populateThrottleAnalyticsDTO and populateFaultAnalyticsDTO. I think you can use a common method for them.

**Pull request merged**